spec.md

```diff
@@ -1090,6 +1084,252 @@ message ControllerServiceCapability {
+  // subsequent calls to refer to the provisioned snapshot.
+  string id = 2;
+
+  // Identity information for the source volume. Note that create
+  // snapshot from snapshot is not supported here so the source has to
+  // be a volume. This field is REQUIRED.
+  string source_volume_id = 3;
+
+  // Timestamp when the point-in-time snapshot is taken on the storage
+  // system. The format of this field should be a UnixNano time encoded
+  // as an int64. This field is REQUIRED.
+  int64 created_at = 4;
+
+  // The status of a snapshot.
+  SnapshotStatus status = 5;
+}
+
+message SnapshotSize {
+  // The disk size of the snapshot. The purpose of this field is to
+  // give CO guidance on how much space is needed to create a volume
+  // from this snapshot. The size of the volume must not be less than
+  // the size of the source snapshot. This field is OPTIONAL. The value
+  // of this field MUST NOT be negative.
+  int64 size_bytes = 1;
+}
+
+// The status of a snapshot. A snapshot could have the following
+// statuses:
+// UPLOADING - A snapshot is cut and is now being uploaded.
+// READY - A snapshot is ready for use.
+// Some cloud providers will upload the snapshot to a location in
+// the cloud (i.e., an object store) after the snapshot is cut.
+// Uploading is a long process that could take hours. At this phase,
+// since the snapshot is already cut, a `thaw` operation can be
+// performed so application can start running again (assuming a
+// `freeze` operation was done before taking the snapshot). The
+// status of the snapshot will become `READY` after the upload is
+// complete.
+// For cloud providers and storage systems that don't have the
+// uploading process, the status should be `READY` after the snapshot
+// is cut. `thaw` can be done when the status is `READY` in this case.
+// A `CREATING` status is not included here because CreateSnapshot
+// is synchronous and will block until the snapshot is cut.
+// The SnapshotStatus parameter provides guidance to the CO on what
+// action can be taken in the process of snapshotting. Based on
+// this information, CO can issue repeated (idemponent) calls to
+// CreateSnapshot, monitor the response, and make decisions.
+// Note that CreateSnapshot is a synchronous call and it must block
+// until the snapshot is cut. If the cloud provider or storage system
+// does not need to upload the snapshot after it is cut, the status
+// returned by CreateSnapshot should be `READY`.
+// If the cloud provider or storage system needs to upload the snapshot
+// after the snapshot is cut, the status returned by CreateSnapshot
+// should be `UPLOADING`. CO can continue to call CreateSnapshot while
+// waiting for the upload to complete until the status becomes `READY`.
+// Note that CreateSnapshot no longer blocks after the snapshot is cut.
+// Alternatively, another RPC ListSnapshots can be called with
+// snapshot_id as filtering repeatedly to wait for the upload to
```

Comment:
a/as/and/ ? maybe "called repeatedly" (earlier in the sentence) instead of "filtering repeatedly" here? maybe it's worth using the term "poll" or "polling" to describe the process we're expecting COs to use here?